File: 133_Clone_Graph.py
# Definition for a Node.
class Node(object):
    def __init__(self, val = 0, neighbors = []):
        self.val = val
        self.neighbors = neighbors

class Solution(object):
    def dfs(self, node, m):
        if not node: return None
        if node in m: return m[node]
        
        nnode = Node(node.val)
        m[node] = nnode
        for neigh in node.neighbors:
            nnode.neighbors.append(self.dfs(neigh, m))
        return nnode
    
    def cloneGraph(self, node):
        """
        :type node: Node
        :rtype: Node
        """
        m = dict()
        return self.dfs(node, m)

File: 133_Clone_Graph.cpp
#include<iostream>
#include<vector>
#include<unordered_map>

using namespace std;

class Node {
public:
    int val;
    vector<Node*> neighbors;
    
    Node() {
        val = 0;
        neighbors = vector<Node*>();
    }
    
    Node(int _val) {
        val = _val;
        neighbors = vector<Node*>();
    }
    
    Node(int _val, vector<Node*> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};

class Solution{
private:
    Node* dfs_copy(Node* node, unordered_map<Node*, Node*>& m){
        if (m.count(node)) return m[node];

        Node* nnode = new Node(node->val);
        m[node] = nnode;
        for (auto neigh : node->neighbors){
            nnode->neighbors.push_back(dfs_copy(neigh, m));
        }
        return nnode;
    }
public:
    Node* cloneGraph(Node* node) {
        if (!node) return NULL;
        unordered_map<Node*, Node*> m;
        return dfs_copy(node, m);
    }
};


int main(){
    return 0;
}